### PR TITLE
Rewrote the dockerfiles to be more extensible and readable

### PR DESCRIPTION
**Pull request by AaronYoung5**
I wasn't necessarily happy with the old dockerfiles. They were pretty confusing to read and they built custom user additions out from bash scripts, which isn't docker-esque. Instead, you'd want to add custom additions to the dockerfile directly so that docker can properly cache earlier changes and reduce build times. However, we don't want to have custom configs in each dockerfile because there would be a lot of duplicate or unnecessary code; for example, device permissions necessary for the nx shouldn't be done in the dev container.

These changes leverage a non-standard library called [`dockerfile-x`](https://codeberg.org/devthefuture/dockerfile-x/). You can call `INCLUDE` to include other dockerfiles, which is then used to replace the existing user scripts. See [this discussion](https://github.com/moby/moby/issues/735) for more information regarding docker support for an include statement.

Open Questions:
- [ ] What needs to be done to support the nx/on-vehicle computers? @harryzhang1018

### CODE REVIEW

**Comment by harryzhang1018** [comment]
You mean what kind of sensor support packages and drivers I need?

**Comment by AaronYoung5** [comment]
> @AaronYoung5 You mean what kind of sensor support packages and drivers I need?

Yes. But I believe this comes after we clean up the NX/ART-1 code. 

Would you wanna test the new chrono container? It now builds chrono::ros. We’ll need to update the simulation demos that we’ve been using.

**Comment by harryzhang1018** [comment]
> > @AaronYoung5 You mean what kind of sensor support packages and drivers I need?
> 
> Yes. But I believe this comes after we clean up the NX/ART-1 code.
> 
> Would you wanna test the new chrono container? It now builds chrono::ros. We’ll need to update the simulation demos that we’ve been using.

@AaronYoung5  Sorry I didn't see this ealier. I will try to test it today about new container setup and lets talk nx related stuffs on that different PR.

**Comment by AaronYoung5** [comment]
@ishaanamahajan added you as a reviewer. Give it a try. You'll need to update ATK to the latest version!

**Review by ishaanamahajan — DISMISSED** [review]

File: docker/snippets/chrono.dockerfile
Comment (harryzhang1018) [inline docker/snippets/chrono.dockerfile:85]: Any specific reason why we wanna delete this? Because sometime, like RC cars, there might be some bugs which prob requires twisting parameters and look into source code.

**Comment by harryzhang1018** [comment]
I have quick suggestion, how about we have both python and cpp script for running simulation demo. For some demos (like RCCars), I felt it's easier to use cpp cuz there are existing demos written in cpp but no python. Felt like most of the demo, chrono makes more detailed cpp than python. Also other future users may want to choose python or cpp by themselves. It won't be too hard I guess, here is link of how to call chrono as library: https://github.com/projectchrono/chrono/tree/feature/ros/template_project

so the structure will more like this: 
1. sim
   1a. demo_py
   1b. demo_cpp

**Comment by AaronYoung5** [comment]
> I have quick suggestion, how about we have both python and cpp script for running simulation demo. For some demos (like RCCars), I felt it's easier to use cpp cuz there are existing demos written in cpp but no python. Felt like most of the demo, chrono makes more detailed cpp than python. Also other future users may want to choose python or cpp by themselves. It won't be too hard I guess, here is link of how to call chrono as library: https://github.com/projectchrono/chrono/tree/feature/ros/template_project
> 
> so the structure will more like this:
> 
> 1. sim
>    1a. demo_py
>    1b. demo_cpp

I mean it's up to you guys. I'd say for fast development, python makes the most sense. The Chrono API is 1:1 identical for the most part with python (minus some custom attributes like ChVectorD, etc.). It's certainly possible to have the cpp, but I'd recommend python for now since we're in flux with our environments and training for ML is always going to be with python.

Agreed? If there's a big reason to use cpp, I can add it, but I think using python makes much more sense.

File: docker/snippets/chrono.dockerfile
Comment (AaronYoung5) [reply docker/snippets/chrono.dockerfile:85]: Hmm I would tree Chrono as strictly a library in this container, as in not editable. How often do you edit the model?

**Comment by ishaanamahajan** [comment]
> > I have quick suggestion, how about we have both python and cpp script for running simulation demo. For some demos (like RCCars), I felt it's easier to use cpp cuz there are existing demos written in cpp but no python. Felt like most of the demo, chrono makes more detailed cpp than python. Also other future users may want to choose python or cpp by themselves. It won't be too hard I guess, here is link of how to call chrono as library: https://github.com/projectchrono/chrono/tree/feature/ros/template_project
> > so the structure will more like this:
> > 
> > 1. sim
> >    1a. demo_py
> >    1b. demo_cpp
> 
> I mean it's up to you guys. I'd say for fast development, python makes the most sense. The Chrono API is 1:1 identical for the most part with python (minus some custom attributes like ChVectorD, etc.). It's certainly possible to have the cpp, but I'd recommend python for now since we're in flux with our environments and training for ML is always going to be with python.
> 
> Agreed? If there's a big reason to use cpp, I can add it, but I think using python makes more sense.

Just a comment - I agree with having python for faster development, but for RCCar stuff, there are a couple of demos only available in .cpp as @harryzhang1018  mentioned. This is something that I have observed while trying to fit in the low-fidelity models into the current stack. If having both of them is doable, and shouldn't be difficult,  having both python and .cpp would really be helpful. Just my two cents.

**Comment by harryzhang1018** [comment]
@AaronYoung5  I don't know if this only happens in my end or @ishaanamahajan  has the same problem as well. The issue is that I can't run `colcon build` on workspace directory.

**Comment by AaronYoung5** [comment]
> @AaronYoung5 I don't know if this only happens in my end or @ishaanamahajan has the same problem as well. The issue is that I can't run `colcon build` on workspace directory.

send error messages if you can!

**Comment by AaronYoung5** [comment]
> > > I have quick suggestion, how about we have both python and cpp script for running simulation demo. For some demos (like RCCars), I felt it's easier to use cpp cuz there are existing demos written in cpp but no python. Felt like most of the demo, chrono makes more detailed cpp than python. Also other future users may want to choose python or cpp by themselves. It won't be too hard I guess, here is link of how to call chrono as library: https://github.com/projectchrono/chrono/tree/feature/ros/template_project
> > > so the structure will more like this:
> > > 
> > > 1. sim
> > >    1a. demo_py
> > >    1b. demo_cpp
> > 
> > 
> > I mean it's up to you guys. I'd say for fast development, python makes the most sense. The Chrono API is 1:1 identical for the most part with python (minus some custom attributes like ChVectorD, etc.). It's certainly possible to have the cpp, but I'd recommend python for now since we're in flux with our environments and training for ML is always going to be with python.
> > Agreed? If there's a big reason to use cpp, I can add it, but I think using python makes more sense.
> 
> Just a comment - I agree with having python for faster development, but for RCCar stuff, there are a couple of demos only available in .cpp as @harryzhang1018 mentioned. This is something that I have observed while trying to fit in the low-fidelity models into the current stack. If having both of them is doable, and shouldn't be difficult, having both python and .cpp would really be helpful. Just my two cents.

Okay, will do! I will do this in a separate PR.

**Comment by AaronYoung5** [comment]
> > @AaronYoung5 I don't know if this only happens in my end or @ishaanamahajan has the same problem as well. The issue is that I can't run `colcon build` on workspace directory.
> 
> send error messages if you can!

@harryzhang1018 pushed fix. try again please :).

**Comment by AaronYoung5** [comment]
So @harryzhang1018 @ishaanamahajan, is this good to merge? Need an approve. I pushed a few commits after ishaan approved. Only approve if you feel it makes sense.

It now doesn't delete chrono so you can edit it in the container, though I wouldn't recommend this since the changes won't persist if you tear down the container.

**Comment by ishaanamahajan** [comment]
Screenshot 2023-10-12 at 10 02 20 AM
@AaronYoung5 I am running into this whenever I try to run a simulation file

**Comment by AaronYoung5** [comment]
> Screenshot 2023-10-12 at 10 02 20 AM @AaronYoung5 I am running into this whenever I try to run a simulation file

I can't replicate. Can you rebuild? `atk dev -dbua -s chrono`

**Comment by ishaanamahajan** [comment]
I rebuilt it, and same issue

**Pull request merged**
